### PR TITLE
Python version support

### PR DESCRIPTION
**Pull request by willtai**
# Description

- Updates  `pr-e2e-tests.yaml` to iterate through different python versions
- Use different `numpy` and `scipy` versions for Python 3.12 to avoid errors when installing the virtual env. This is how [`langchain_huggingface` handles it](https://github.com/langchain-ai/langchain/blob/master/libs/partners/huggingface/pyproject.toml#L31) 
- Python 3.8 requires `typing.List`, `typing.Dict`, and `typing.Tuple` to be used instead of `list`, `dict`, and `tuple`.
- Resolve mypy errors

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Documentation update
- [x] Project configuration change

## Complexity

Complexity: Low

## How Has This Been Tested?
- [ ] Unit tests
- [x] E2E tests
- [x] Manual tests

[This was tested successfully on my fork's PR](https://github.com/willtai/neo4j-genai-python/pull/1).

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [x] Documentation has been updated
- [ ] Unit tests have been updated
- [x] E2E tests have been updated
- [x] Examples have been updated
- [ ] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed
- [ ] CHANGELOG.md updated if appropriate

### CODE REVIEW

File: tests/e2e/weaviate_e2e/test_weaviate_e2e.py [hunk 1]
```diff
@@ -22,7 +22,7 @@
 from neo4j_genai.embedder import Embedder
 from neo4j_genai.retrievers import WeaviateNeo4jRetriever
 from neo4j_genai.types import RetrieverResult, RetrieverResultItem
-from weaviate.client import Client
+from weaviate.client import WeaviateClient
```
Comment (willtai) [inline tests/e2e/weaviate_e2e/test_weaviate_e2e.py:25]: Moved to using WeaviateClient due to v4 migration, see

https://weaviate.io/developers/weaviate/client-libraries/python/v3_v4_migration
https://github.com/run-llama/llama_index/pull/13719/files
This was causing mypy errors

File: pyproject.toml [hunk 2]
```diff
@@ -50,13 +50,21 @@ langchain-community = "^0.2.0"
 requests = "^2.32.0"
 sphinx = { version = "^7.2.6", python = "^3.9" }
 langchain-huggingface = "^0.0.3"
+tox = "^4.15.1"
+numpy = [
+    {version = "^1.24.0", python = "<3.12"},
```
Comment (stellasia) [inline pyproject.toml:55]: I didn't know we could specify versions like this, nice!

File: tests/unit/retrievers/test_base.py [hunk 3]
```diff
@@ -12,12 +12,14 @@
 #  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
+from __future__ import annotations  # Reminder: May be removed after Python 3.9 is EOL.
```
Comment (stellasia) [inline tests/unit/retrievers/test_base.py:15]: Out of curiosity, why is this import required only in this file?

File: .github/workflows/pr-e2e-tests.yaml [hunk 4]
```diff
@@ -13,6 +15,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
+        python-version: ['3.8', '3.9', '3.10', '3.11', '3.12']
```
Comment (stellasia) [inline .github/workflows/pr-e2e-tests.yaml:18]: We do not want to do the same for unit tests?

File: tests/unit/retrievers/test_base.py (hunk 3 above)
Comment (willtai) [reply tests/unit/retrievers/test_base.py:15]: We could do it for every file too tbh and we can drop it everywhere if we stop supporting Python 3.8 in future

File: .github/workflows/pr-e2e-tests.yaml (hunk 4 above)
Comment (willtai) [reply .github/workflows/pr-e2e-tests.yaml:18]: Added in https://github.com/neo4j/neo4j-genai-python/pull/67/commits/43e1cf56cf7fd1920693fda75f5c25f798f38a03

File: tests/unit/retrievers/test_base.py (hunk 3 above)
Comment (willtai) [reply tests/unit/retrievers/test_base.py:15]: https://github.com/neo4j/neo4j-genai-python/pull/67/commits/46670c0fee52b942ff18a50c13b247b48e269595

**Review by stellasia — APPROVED** [review]
Thanks, very helpful!

**Comment by willtai** [comment]
@CodiumAI-Agent /update_changelog

**Pull request merged**